test/src/utils/quaternion.test.cpp

```diff
@@ -0,0 +1,353 @@
+ * fixed size of the array generated as that is the fixed size of
+ * a quaternion instance.
+ * @returns array of 4 integers
+ */
+std::array<int, 4> createRandomArrayForQuaternion()
+{
+  std::srand(time(0));
+  std::array<int, 4> output = std::array<int, 4>();
+  for (int i = 0;i < 4;i++) {
+    output[i] = rand()%1000;
+  }
+  return output;
+}
+
+/**
+ * @brief Function which compares two floats; code taken from:
+ * https://www.tutorialspoint.com/floating-point-comparison-in-cplusplus
+ * @param elem1 float number to be compared
+ * @param elem2 float number to be comddddpared
+ * @param epsilon
+ * @returns true if floats to be compared are equal
+ */
+bool compare_float_for_quaternion(float elem1, float elem2, float epsilon = 0.1f)
+{
+  if (fabs(elem1 - elem2) < epsilon) {
+       return true;
+  } else {
+       return false;
+  }
+}
+
+/**
+ * @brief Function which calculates the norm of a quaternion
+ * @param Quaternion instance
+ * @returns float containing the computed norm of the quaternion.
+ */
+float calculateNorm(Quaternion<int> quat)
+{
+  float sumSquare = 0;
+  for (int i = 0; i < 4; i++) {
+    sumSquare += pow(quat[i], 2);
+  }
+  return sqrt(sumSquare);
+}
+
+// -------------------------------------------------------------------------------------------------
+// Functionality
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used for testing the functionality
+ * of quaternion.hpp
+ */
+struct QuaternionFunctionality : public::testing::Test
+{
+    /* data */
+    const int kSizeOfQuaternion = 4;
+    Quaternion<int> quaternion_one;
+    Quaternion<int> quaternion_two;
+
+    Quaternion<int> quaternion_result_one;
+    Quaternion<int> quaternion_result_two;
+
+    Vector<int, 4> test_vector;
+    Vector<int, 3> test_vector_3d;
+
+    float arithmetic_result;
+
+    int kValue;
+
+    void SetUp()
+    {
+        kValue = rand()%1000;
+        quaternion_one = Quaternion<int>(rand()%1000, rand()%1000, rand()%1000, rand()%1000);
+        quaternion_two = Quaternion<int>(rand()%1000, rand()%1000, rand()%1000, rand()%1000);
+    }
+};
+
+// -------------------------------------------------------------------------------------------------
+// Tests
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * @brief Test to determine the correct instantiation of the zero
+ * Quaternion class instance
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionOfZeroQuaternion)
+{
+    Quaternion<int> zeroQuaternion = Quaternion<int>(0);
+    ASSERT_EQ(0, zeroQuaternion.norm());
+    for (int i = 0; i < kSizeOfQuaternion; i++) {
+        ASSERT_EQ(0, zeroQuaternion[i]);
+    }
+}
+
+/**
+ * @brief Test to determine the correct instantiation of a non zero
+ * Quaternion class instance.
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionOfNonZeroQuaternion)
```

Comment:
Maybe instead of just one Testsuite, it would be helpful to divide them into categories, also If you don't need a struct you can use TEST(name_of_the_test_suite,name_of_the_test)

Reply:
Maybe a test suite for the constructor checks

Reply:
Would you recommend having two separate files for that, just like with the vector tests? Or just one file?

Reply:
I don't think that you have so many so it is not needed